# Data Quality Assessment

Pada tahap ini dilakukan pemeriksaan terhadap setiap dataset untuk mengetahui:

- Jumlah baris dan kolom
- Tipe data setiap kolom
- Missing values
- Duplikasi data

## Objectives

The objectives of this notebook are:

- Load every dataset
- Understand each table
- Identify data types
- Detect missing values
- Detect duplicate records
- Understand relationships between datasets

In [28]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

In [29]:
customers = pd.read_csv("../data/raw/olist_customers_dataset.csv")
orders = pd.read_csv("../data/raw/olist_orders_dataset.csv")
order_items = pd.read_csv("../data/raw/olist_order_items_dataset.csv")
payments = pd.read_csv("../data/raw/olist_order_payments_dataset.csv")
products = pd.read_csv("../data/raw/olist_products_dataset.csv")
reviews = pd.read_csv("../data/raw/olist_order_reviews_dataset.csv")
sellers = pd.read_csv("../data/raw/olist_sellers_dataset.csv")
geolocation = pd.read_csv("../data/raw/olist_geolocation_dataset.csv")
category = pd.read_csv("../data/raw/product_category_name_translation.csv")

In [30]:
datasets = {
    "Customers": customers,
    "Orders": orders,
    "Order Items": order_items,
    "Payments": payments,
    "Products": products,
    "Reviews": reviews,
    "Sellers": sellers,
    "Geolocation": geolocation,
    "Category Translation": category
}

In [31]:
overview = []

for name, df in datasets.items():
    overview.append({
        "Dataset": name,
        "Rows": df.shape[0],
        "Columns": df.shape[1]
    })

overview = pd.DataFrame(overview)
overview

,Dataset,Rows,Columns
0,Customers,99441,5
1,Orders,99441,8
2,Order Items,112650,7
3,Payments,103886,5
4,Products,32951,9
5,Reviews,99224,7
6,Sellers,3095,4
7,Geolocation,1000163,5
8,Category Translation,71,2


In [32]:
customers = pd.read_csv("../data/raw/olist_customers_dataset.csv")
customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [33]:
customers = pd.read_csv("../data/raw/olist_customers_dataset.csv")

orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [34]:
for name, df in datasets.items():

    print("="*60)
    print(name)
    display(df.dtypes)

Customers


customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

Orders


order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

Order Items


order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object

Payments


order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64
dtype: object

Products


product_id                     object
product_category_name          object
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object

Reviews


review_id                  object
order_id                   object
review_score                int64
review_comment_title       object
review_comment_message     object
review_creation_date       object
review_answer_timestamp    object
dtype: object

Sellers


seller_id                 object
seller_zip_code_prefix     int64
seller_city               object
seller_state              object
dtype: object

Geolocation


geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                object
geolocation_state               object
dtype: object

Category Translation


product_category_name            object
product_category_name_english    object
dtype: object

In [35]:
customers.describe(include="all")

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
count,99441,99441,99441.000000,99441,99441
unique,99441,96096,NaN,4119,27
top,274fa6071e5e17fe303b9748641082c8,8d50f5eadf50201ccdcedfb9e2ac8455,NaN,sao paulo,SP
freq,1,17,NaN,15540,41746
mean,NaN,NaN,35137.474583,NaN,NaN
std,NaN,NaN,29797.938996,NaN,NaN
min,NaN,NaN,1003.000000,NaN,NaN
25%,NaN,NaN,11347.000000,NaN,NaN
50%,NaN,NaN,24416.000000,NaN,NaN
75%,NaN,NaN,58900.000000,NaN,NaN


In [36]:
missing_summary = []

for name, df in datasets.items():

    missing = df.isnull().sum()

    percent = (missing/len(df)*100).round(2)

    for col in df.columns:

        missing_summary.append({

            "Dataset": name,

            "Column": col,

            "Missing": missing[col],

            "Percentage": percent[col]

        })

missing_summary = pd.DataFrame(missing_summary)

missing_summary.sort_values(
    "Missing",
    ascending=False
)

,Dataset,Column,Missing,Percentage
37,Reviews,review_comment_title,87656,88.34
38,Reviews,review_comment_message,58247,58.70
11,Orders,order_delivered_customer_date,2965,2.98
10,Orders,order_delivered_carrier_date,1783,1.79
27,Products,product_name_lenght,610,1.85
26,Products,product_category_name,610,1.85
28,Products,product_description_lenght,610,1.85
29,Products,product_photos_qty,610,1.85
9,Orders,order_approved_at,160,0.16
31,Products,product_length_cm,2,0.01


In [37]:
duplicates = []

for name, df in datasets.items():

    duplicates.append({

        "Dataset": name,

        "Duplicate Rows": df.duplicated().sum()

    })

duplicates = pd.DataFrame(duplicates)

duplicates

,Dataset,Duplicate Rows
0,Customers,0
1,Orders,0
2,Order Items,0
3,Payments,0
4,Products,0
5,Reviews,0
6,Sellers,0
7,Geolocation,261831
8,Category Translation,0


# Initial Findings

## 1. Dataset Overview

The Olist Brazilian E-commerce dataset consists of **9 relational tables** containing information about customers, orders, products, sellers, payments, reviews, geolocation, and product category translations.

| Dataset | Rows | Columns |
|---------|------:|--------:|
| Customers | 99,441 | 5 |
| Orders | 99,441 | 8 |
| Order Items | 112,650 | 7 |
| Payments | 103,886 | 5 |
| Products | 32,951 | 9 |
| Reviews | 99,224 | 7 |
| Sellers | 3,095 | 4 |
| Geolocation | 1,000,163 | 5 |
| Category Translation | 71 | 2 |

The **Orders** table acts as the central dataset because it connects customers, products, payments, sellers, and reviews.

## 2. Data Quality Assessment

Based on the initial inspection, several data quality issues were identified.

### Missing Values

The following columns contain missing values:

| Dataset | Column | Missing |
|---------|---------|--------:|
| Reviews | review_comment_title | 87,656 |
| Reviews | review_comment_message | 58,247 |
| Orders | order_delivered_customer_date | 2,965 |
| Orders | order_delivered_carrier_date | 1,783 |
| Products | Product attributes | 610 |

Most missing values occur in the review comment columns because not every customer leaves textual feedback after completing an order.

The missing delivery timestamps are likely associated with orders that were canceled or had not yet been delivered.

Product-related missing values represent only a small percentage (approximately **1.85%**) of the dataset.

## 3. Duplicate Records

Duplicate record inspection shows that almost every dataset does not contain duplicated rows.

However, the **Geolocation** dataset contains **261,831 duplicated records**.

These duplicates are expected because multiple customers may share the same ZIP code and geographic coordinates.

The necessity of removing these duplicates will be evaluated during the data cleaning stage.

## 4. Data Structure

The datasets follow a relational database structure.

Main relationships include:

- Customers → Orders
- Orders → Order Items
- Order Items → Products
- Order Items → Sellers
- Orders → Payments
- Orders → Reviews
- Products → Category Translation

This structure allows multiple business analyses such as sales performance, customer behavior, product analysis, seller performance, and logistics.

## 5. Data Readiness

Overall, the dataset is well structured and suitable for analysis.

Before conducting exploratory data analysis, several preprocessing tasks are required:

- Convert timestamp columns into datetime format.
- Handle missing values.
- Remove unnecessary duplicate records (if applicable).
- Merge related tables into a unified analytical dataset.

## Next Step

The next notebook (**02_data_cleaning.ipynb**) will focus on preparing the datasets by performing data cleaning, handling missing values, converting data types, and merging the required tables for analysis.